<a href="https://colab.research.google.com/github/IvlevaAnna1/Hometasks/blob/main/%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_%D0%BA_%D0%B7%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D1%8E_4_Ivleva.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практика
Используемые библиотеки

In [ ]:
import pandas as pd
import numpy as np
import requests # for web-download
import io # for web-download
import re # for data processing


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Загрузка DataFrame
### Задача 1
На основании данных портала "Открытые данные России" о результатах Химического анализа родника в Нескучном саду https://data.gov.ru/opendata/7708660670-rodnik-neskuchniy-sad
средствами библиотеки __Pandas__ сформируйте поле выводов по каждому анализирумомому параметру.
Например, по показателю _pH_ получен результат _8.4 единицы pH_ при нормативе от _6 до 9 единиц pH_. Т.о. по данному показателю результат анализа в норме.
Для решения задачи необходимо программно "прочитать и понять" значение столбца "Норматив" и выделенное численное значение сравнить с нормативом согласно логике норматива. Например, __6 >= pH >= 9__.
В итоговом DataFrame столбец "Показатель" сделайте индексным.


Загзрузка DataFrame выполняется непосредственно c сайта "Открытые данные России" https://data.gov.ru/opendata/7708660670-rodnik-neskuchniy-sad/data-20160608T1215-structure-20160608T1215.csv (см. код ниже).


In [231]:
# headers={
#     'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0',
# }
# def get_content(url):
#     with requests.Session() as req:
#         req.headers.update(headers)
#         r = req.get(url).content
#     return r

# url ="https://data.gov.ru/opendata/7708660670-rodnik-neskuchniy-sad/data-20160608T1215-structure-20160608T1215.csv"
# s = get_content(url)
# df=pd.read_csv(io.StringIO(s.decode('UTF8')))


### Если не работает загрузка on-line
df=pd.read_csv('/content/drive/MyDrive/Him_analiz.csv', sep=';', index_col=0)
display(df)

Единица измерений  Результат анализа  \
Показатель                                                                   
pH                                           единицы pH                8.4   
Запах                                             баллы                  1   
Цветность                                       градусы            б/цвета   
Жёсткость                                   мг-эквл/дм3  9.199999999999999   
Аммиак и аммоний-ион (по азоту)                  мг/дм3               0.42   
Нитриты (по NO2)                                 мг/дм3              0.017   
Нитраты (по NO3)                                 мг/дм3                 24   
Фосфаты (P)                                      мг/дм3               0.36   
Хлориды (Cl)                                     мг/дм3                200   
Сульфаты (SO4)                                   мг/дм3              189.5   
Железо (включая хлорное железо) по Fe            мг/дм3              0.019   
Нефть                                            мг/дм3               0.55   
Общая минерализация (сухой остаток)              мг/дм3                590   
Окисляемость перманганатная                      мг/дм3                  2   

                                              Норматив  
Показатель                                              
pH                                      в пределах 6-9  
Запах                                     не более 2-3  
Цветность                                  не более 30  
Жёсткость                              в пределах 7-10  
Аммиак и аммоний-ион (по азоту)           не более 1,5  
Нитриты (по NO2)                          не более 3,3  
Нитраты (по NO3)                           не более 45  
Фосфаты (P)                               не более 3,5  
Хлориды (Cl)                              не более 350  
Сульфаты (SO4)                            не более 500  
Железо (включая хлорное железо) по Fe     не более 0,3  
Нефть                                     не более 0,3  
Общая минерализация (сухой остаток)      не более 1000  
Окисляемость перманганатная               не более 0,5

In [232]:
# Ваше решение
df.rename(columns={'Показатель': 'Indicator','Единица измерений': 'Unit','Результат анализа': 'Result','Норматив': 'Norma'}, inplace=True)
df['Result'] = df['Result'].replace('б/цвета', '0')
df['Norma'] = df['Norma'].str.replace(',', '.')
df

Unit             Result  \
Показатель                                                              
pH                                      единицы pH                8.4   
Запах                                        баллы                  1   
Цветность                                  градусы                  0   
Жёсткость                              мг-эквл/дм3  9.199999999999999   
Аммиак и аммоний-ион (по азоту)             мг/дм3               0.42   
Нитриты (по NO2)                            мг/дм3              0.017   
Нитраты (по NO3)                            мг/дм3                 24   
Фосфаты (P)                                 мг/дм3               0.36   
Хлориды (Cl)                                мг/дм3                200   
Сульфаты (SO4)                              мг/дм3              189.5   
Железо (включая хлорное железо) по Fe       мг/дм3              0.019   
Нефть                                       мг/дм3               0.55   
Общая минерализация (сухой остаток)         мг/дм3                590   
Окисляемость перманганатная                 мг/дм3                  2   

                                                 Norma  
Показатель                                              
pH                                      в пределах 6-9  
Запах                                     не более 2-3  
Цветность                                  не более 30  
Жёсткость                              в пределах 7-10  
Аммиак и аммоний-ион (по азоту)           не более 1.5  
Нитриты (по NO2)                          не более 3.3  
Нитраты (по NO3)                           не более 45  
Фосфаты (P)                               не более 3.5  
Хлориды (Cl)                              не более 350  
Сульфаты (SO4)                            не более 500  
Железо (включая хлорное железо) по Fe     не более 0.3  
Нефть                                     не более 0.3  
Общая минерализация (сухой остаток)      не более 1000  
Окисляемость перманганатная               не более 0.5

In [233]:
list = []
for index, row in df.iterrows():
  if "в пределах" in row['Norma']:
    if float(str(row['Norma'].replace('в пределах ','').split('-')[0])) <= float(str(row['Result'])) <= float(str(row['Norma'].replace('в пределах ','').split('-')[1])):
      list.append("Показатель в норме")
    elif float(str(row['Norma'].replace('в пределах ','').split('-')[0])) > float(str(row['Result'])):
      list.append("Показатель ниже нормы")
    elif float(str(row['Norma'].replace('в пределах ','').split('-')[1])) > float(str(row['Result'])):
      list.append("Показатель выше нормы")
  elif "не более " in row['Norma']:
    if float(str(row['Norma'].replace('не более ','').split('-')[-1])) >= float(str(row['Result'])):
      list.append("Показатель в норме")
    elif float(str(row['Norma'].replace('не более ',''))) < float(str(row['Result'])):
      list.append("Показатель выше нормы")

In [234]:
df.insert(loc= len(df.columns) , column='Description', value=list)

In [235]:
df

Unit             Result  \
Показатель                                                              
pH                                      единицы pH                8.4   
Запах                                        баллы                  1   
Цветность                                  градусы                  0   
Жёсткость                              мг-эквл/дм3  9.199999999999999   
Аммиак и аммоний-ион (по азоту)             мг/дм3               0.42   
Нитриты (по NO2)                            мг/дм3              0.017   
Нитраты (по NO3)                            мг/дм3                 24   
Фосфаты (P)                                 мг/дм3               0.36   
Хлориды (Cl)                                мг/дм3                200   
Сульфаты (SO4)                              мг/дм3              189.5   
Железо (включая хлорное железо) по Fe       мг/дм3              0.019   
Нефть                                       мг/дм3               0.55   
Общая минерализация (сухой остаток)         мг/дм3                590   
Окисляемость перманганатная                 мг/дм3                  2   

                                                 Norma            Description  
Показатель                                                                     
pH                                      в пределах 6-9     Показатель в норме  
Запах                                     не более 2-3     Показатель в норме  
Цветность                                  не более 30     Показатель в норме  
Жёсткость                              в пределах 7-10     Показатель в норме  
Аммиак и аммоний-ион (по азоту)           не более 1.5     Показатель в норме  
Нитриты (по NO2)                          не более 3.3     Показатель в норме  
Нитраты (по NO3)                           не более 45     Показатель в норме  
Фосфаты (P)                               не более 3.5     Показатель в норме  
Хлориды (Cl)                              не более 350     Показатель в норме  
Сульфаты (SO4)                            не более 500     Показатель в норме  
Железо (включая хлорное железо) по Fe     не более 0.3     Показатель в норме  
Нефть                                     не более 0.3  Показатель выше нормы  
Общая минерализация (сухой остаток)      не более 1000     Показатель в норме  
Окисляемость перманганатная               не более 0.5  Показатель выше нормы

## Теория вероятности. События

Требуется сгенерировать необходимые выборки и произвести по ним расчеты

### Задача 2
В ящике 5 апельсинов и 4 яблока. Наудачу выбираются 3 фрукта. Какова вероятность, что все три фрукта – апельсины?

В интернете полученный аналитически ответ 0.119. Подтверждается ли он эксперементально?


In [190]:
# Ваше решение
# Полученный аналитически ответ 0.119 подтверждается экспериментально при наличии достаточного количества наблюдений
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import random
import scipy.stats as sts
%matplotlib inline
from itertools import product

import warnings
warnings.filterwarnings('ignore')

# заполняем ящик
box = ["orange"]*5 + ['apple']*4

# задаем число экспериментов
n = 1000000

# выборка
k = 3

# считаем вероятность
positive = 0
for _ in range(n):
    random.shuffle(box)
    result = all(x == 'orange' for x in box[:k])
    positive += result
print(f'Вероятность вытащить три апельсина {positive/n}')

Вероятность вытащить три апельсина 0.119093


Полученный аналитически ответ 0.119 подтверждается экспериментально при наличии достаточного количества наблюдений

### Задача 3
Мастер, имея 10 деталей, из которых 3 – нестандартных, проверяет детали одну за другой, пока ему не попадется стандартная. Какова вероятность, что он проверит ровно две детали?


В интернете полученный аналитически ответ 7/30 или 0.23333. Подтверждается ли он эксперементально?

In [230]:
# Ваше решение
# список деталей
details = ['non_standart']*3 + ["standart"]*7

# задаем число экспериментов
n = 1000000

# выборка
k = 2

# считаем вероятность
positive = 0
for _ in range(n):
    #перемешиваем исходный список с деталями
    random.shuffle(details)
    #если первая выбранная деталь является нестандартной, то condition1 будет равно True
    condition1 = all(x == 'non_standart' for x in details[0:k-1])
    #если вторая выбранная деталь является стандартной, то condition2 будет равно True
    condition2 = all(x == 'standart' for x in details[k-1:k])
    #результирующее условие должно давать True в случае выполнения condition1 и condition2
    result = all([condition1, condition2])
    positive += result
print(f'Вероятность вытащить стандартную деталь со второй попытки {positive/n}')

Вероятность вытащить стандартную деталь со второй попытки 0.233112


Полученный аналитически ответ 0.23 подтверждается экспериментально при наличии достаточного количества наблюдений